src/arrayvec_copy.rs

```diff
@@ -0,0 +1,1218 @@
+/// }
+/// ```
+impl<'a, T: Copy + 'a, const CAP: usize> IntoIterator for &'a mut ArrayVecCopy<T, CAP> {
+    type Item = &'a mut T;
+    type IntoIter = slice::IterMut<'a, T>;
+    fn into_iter(self) -> Self::IntoIter { self.iter_mut() }
+}
+
+/// Iterate the `ArrayVecCopy` with each element by value.
+///
+/// The vector is consumed by this operation.
+///
+/// ```
+/// use arrayvec::ArrayVecCopy;
+///
+/// for elt in ArrayVecCopy::from([1, 2, 3]) {
+///     // ...
+/// }
+/// ```
+impl<T: Copy, const CAP: usize> IntoIterator for ArrayVecCopy<T, CAP> {
+    type Item = T;
+    type IntoIter = IntoIter<T, CAP>;
+    fn into_iter(self) -> IntoIter<T, CAP> {
+        IntoIter { index: 0, v: self, }
+    }
+}
+
+
+/// By-value iterator for `ArrayVecCopy`.
+pub struct IntoIter<T: Copy, const CAP: usize> {
+    index: usize,
+    v: ArrayVecCopy<T, CAP>,
+}
+
+impl<T: Copy, const CAP: usize> Iterator for IntoIter<T, CAP> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index == self.v.len() {
+            None
+        } else {
+            unsafe {
+                let index = self.index;
+                self.index = index + 1;
+                Some(ptr::read(self.v.get_unchecked_ptr(index)))
+            }
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.v.len() - self.index;
+        (len, Some(len))
+    }
+}
+
+impl<T: Copy, const CAP: usize> DoubleEndedIterator for IntoIter<T, CAP> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if self.index == self.v.len() {
+            None
+        } else {
+            unsafe {
+                let new_len = self.v.len() - 1;
+                self.v.set_len(new_len);
+                Some(ptr::read(self.v.get_unchecked_ptr(new_len)))
+            }
+        }
+    }
+}
+
+impl<T: Copy, const CAP: usize> ExactSizeIterator for IntoIter<T, CAP> { }
+
+impl<T: Copy, const CAP: usize> Drop for IntoIter<T, CAP> {
+    fn drop(&mut self) {
+        // panic safety: Set length to 0 before dropping elements.
+        let index = self.index;
+        let len = self.v.len();
+        unsafe {
+            self.v.set_len(0);
+            let elements = slice::from_raw_parts_mut(
+                self.v.get_unchecked_ptr(index),
+                len - index);
+            ptr::drop_in_place(elements);
+        }
+    }
+}
+
+impl<T: Copy, const CAP: usize> Clone for IntoIter<T, CAP> {
+    fn clone(&self) -> IntoIter<T, CAP> {
+        let mut v = ArrayVecCopy::new();
+        v.extend_from_slice(&self.v[self.index..]);
+        v.into_iter()
+    }
+}
+
+impl<T: Copy, const CAP: usize> fmt::Debug for IntoIter<T, CAP>
+where
+    T: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_list()
+            .entries(&self.v[self.index..])
+            .finish()
+    }
+}
+
+/// A draining iterator for `ArrayVecCopy`.
+pub struct Drain<'a, T: Copy + 'a, const CAP: usize> {
```

Comment:
Drain could be a candidate for sharing between the two implementations, even if it requires a bit of refactoring.